### PR TITLE
examples: Cleanups/fixes for docker deps and builds

### LINKED ISSUE #20915: Can not build for front-proxy of examples
Can't build front-proxy example with docker-compose.
It has been failed when pulling python images used in shared resources (flask and tracing).
And I have solved to change the image from 3.10-slim-bullseye to 3.11.0a7-bullseye

I think it could be a docker registry problem rather than the example.
I'll register the issue just in case.
And the detailed error log is below.

=> ERROR [internal] load metadata for docker.io/library/flask_service:python-3.10-slim-bullseye                                                                                                          
------
 > [internal] load metadata for docker.io/library/flask_service:python-3.10-slim-bullseye:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorizatio
ERROR: Service 'tracing' failed to build : Build failed

### PR DESCRIPTION
**Pull request by phlax**
- Pin all docker image deps
- Add dependabot monitoring for docker deps
- Make ports dynamic so examples can be run (by CI) simultaneously
- Remove unused exposed ports
- Quieten apt/pip/service logging
- Make `docker-compose` bin path configurable
- Fix #20915 
- Fix flakey websockets example (hopefully)

Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: examples/cors/backend/docker-compose.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 version: "3.8"
 services:
 
-  flask_service:
+  000-flask_service:
```
Comment (phlax) [inline examples/cors/backend/docker-compose.yaml:4]: dont do this here/now

File: examples/csrf/samesite/Dockerfile-service [hunk 2]
```diff
@@ -0,0 +1 @@
+FROM flask_service:python-3.10-slim-bullseye
```
Comment (phlax) [inline examples/csrf/samesite/Dockerfile-service:1]: make this shared

File: examples/ext_authz/Dockerfile-service [hunk 3]
```diff
@@ -0,0 +1 @@
+FROM flask_service:python-3.10-slim-bullseye
```
Comment (phlax) [inline examples/ext_authz/Dockerfile-service:1]: shared

File: examples/front-proxy/docker-compose.yaml [hunk 4]
```diff
@@ -2,15 +2,15 @@ version: "3.8"
 services:
 
   # base images
-  flask_service:
+  000-flask_service:
```
Comment (phlax) [inline examples/front-proxy/docker-compose.yaml:5]: revert

File: examples/load-reporting-service/Dockerfile-service [hunk 5]
```diff
@@ -0,0 +1 @@
+FROM flask_service:python-3.10-slim-bullseye
```
Comment (phlax) [inline examples/load-reporting-service/Dockerfile-service:1]: shared

File: examples/locality-load-balancing/Dockerfile-service [hunk 6]
```diff
@@ -0,0 +1 @@
+FROM flask_service:python-3.10-slim-bullseye
```
Comment (phlax) [inline examples/locality-load-balancing/Dockerfile-service:1]: shared

File: examples/udp/Dockerfile-udp [hunk 7]
```diff
@@ -0,0 +1 @@
+FROM mendhak/udp-listener
```
Comment (phlax) [inline examples/udp/Dockerfile-udp:1]: pin

File: examples/websocket/docker-compose.yaml [hunk 8]
```diff
@@ -6,29 +6,39 @@ services:
...
 
   proxy-wss-passthrough:
     build:
       context: .
       dockerfile: Dockerfile-proxy-wss-passthrough
     ports:
-    - "30000:10000"
+    - "${PORT_PROXY2:-30000}:10000"
+
+  client-ws:
```
Comment (phlax) [inline examples/websocket/docker-compose.yaml:25]: prevent this from auto starting

File: examples/jaeger-tracing/docker-compose.yaml [hunk 9]
```diff
@@ -24,8 +24,7 @@ services:
       context: .
       dockerfile: Dockerfile-frontenvoy
     ports:
-    - "8000:8000"
-    - "8001:8001"
+    - "${PORT_PROXY:-10000}:8000"
 
   service1:
     image: envoyproxy:tracing
```
Comment (phlax) [inline examples/jaeger-tracing/docker-compose.yaml:30]: fix these

File: examples/front-proxy/verify.sh [hunk 10]
```diff
@@ -1,44 +1,46 @@
 #!/bin/bash -e
 
 export NAME=front-proxy
+export PORT_PROXY="${FRONT_PROXY_PORT_PROXY:-10600}"
+export PORT_HTTPS="${FRONT_PROXY_PORT_PROXY:-10601}"
```
Comment (phlax) [inline examples/front-proxy/verify.sh:5]:
```suggestion
export PORT_HTTPS="${FRONT_PROXY_PORT_HTTPS:-10601}"
```

similar below

File: examples/gzip/verify.sh [hunk 11]
```diff
@@ -1,30 +1,33 @@
 #!/bin/bash -e
 
 export NAME=gzip
+export PORT_PROXY="${GZIP_PORT_PROXY:-10700}"
+export PORT_STATS0="${GZIP_PORT_PROXY:-10701}"
```
Comment (phlax) [inline examples/gzip/verify.sh:5]: similar env var issue

File: examples/jaeger-native-tracing/verify.sh [hunk 12]
```diff
@@ -2,6 +2,8 @@
 
 export NAME=jaeger-native
 export DELAY=10
+export PORT_PROXY="${JAEGER_NATIVE_PORT_PROXY:-11000}"
+export PORT_UI="${JAEGER_NATIVE_PORT_PROXY:-11001}"
```
Comment (phlax) [inline examples/jaeger-native-tracing/verify.sh:6]: and again 8/

File: examples/websocket/interact.sh [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+interact_ws () {
+    local port="$1" \
+          protocol="$2" \
+          backend="$3" \
+          insecure=""
+    if [ "$protocol" == "wss" ]; then
+        insecure="--insecure"
+    fi
+    # echo "websocat $insecure $protocol://$server:10000"
```
Comment (phlax) [inline examples/websocket/interact.sh:11]: rm

**Review by wrowe — APPROVED** [review]
This looks great, I have one question to resolve first.

What is the philosophy on updating these newly pinned dependencies as they bit rot? Is this something that needs to be documented?

**Comment by wrowe** [comment]
The same question as before, though, before we merge, is there a logic to keeping the pinned dependencies from bit-rotting, @phlax?

**Comment by phlax** [comment]
> The same question as before,...

Similar answer to before - there is `examples/DEVELOPER.md` which is my intention to update - but i have a few more changes i would like to make first

this one is a pretty straight refactoring - it shouldnt change how things work - but as you rightly point out it changes the standards for including deps amongst other things - im defo happy to follow if that is ok

**Comment by wrowe** [comment]
That's great, please create an issue pointing to these two (and other relevant) PR's to note the WIP/TODO?

**Comment by Moonsilence** [comment]
Hi! Came here from #20915 which is supposedly fixed with this here thread. But I still get `ERROR [front-proxy_service2 internal] load metadata for docker.io/library/flask_service:python-3.10-slim-bullseye`
That image doesn't exits in the repo nor locally.

**Comment by Moonsilence** [comment]
Hi @phlax !

I cloned the repo this morning.

git rev-parse HEAD
cedb9885e3732a20ce43032f4a6778913a0dd941

**Comment by phlax** [comment]
can you post the exact commands you are running when this happens

also, maybe docker/docker-compose versions

**Comment by Moonsilence** [comment]
Sure, appreciated!

I am trying to run the Sandbox "Front Proxy" from the latest official Envoy documentation:
https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/front_proxy#step-1-start-all-of-our-containers

The very first command is:
docker-compose build --pull

My version of docker-compose: v2.6.1

Error Messages:

` => CANCELED [front-proxy_front-envoy internal] load metadata for docker.io/envoyproxy/envoy-dev:latest                                                               2.0s
 => CANCELED [flask_service:python-3.10-slim-bullseye internal] load metadata for docker.io/library/python:3.10.5-slim-bullseye@sha256:2ae2b820fbcf4e1c5354ec39d0c7e  2.0s
 => ERROR [front-proxy_service1 internal] load metadata for docker.io/library/flask_service:python-3.10-slim-bullseye                                                 2.0s
------
 > [front-proxy_service1 internal] load metadata for docker.io/library/flask_service:python-3.10-slim-bullseye:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed`

**Comment by phlax** [comment]
ok - i see

some of the docs have the instruction of `docker-compose up --build -d` this is what i tested when fixing

and i think should work with little difference in outcome

ill update the docs to reflect the better command

can you confirm _this_ command works ok @Moonsilence ?

**Comment by phlax** [comment]
hmm, actually im realising that doesnt pull the images if they exist already and that this was added for a reason (albeit an imperfect one)

ill test again further - i was convinced i fixed this at least for sme of the examples, but now im wondering...

**Comment by phlax** [comment]
ok  - so it seems my fix didnt work 8( - altho it was needed for other reasons (and i think partially resolved the issue)

what does seem to work is doing the following...

```console

$ docker-compose pull
$ docker-compose up --build -d

```

ill update the docs with these instructions

```

**Comment by phlax** [comment]
hmm - thats odd this is the outpuf of pull for me

```console

..front-proxy$ docker-compose pull
Pulling flask_service ... done
Pulling front-envoy   ... done
Pulling service1      ... done
Pulling service2      ... done
WARNING: Some service image(s) must be built from source by running:
    docker-compose build flask_service
..front-proxy$ echo "$?"
0
```

```console

$ docker --version
Docker version 20.10.10, build b485636

$ docker-compose --version
docker-compose version 1.29.2, build unknown

```

ill update the docs on the basis that it works mostly (i have tested in other similar env - debian <> ubuntu)

not sure why you are getting an error still - but v2.6.1 doesnt seem to exist for docker-compose

**Comment by phlax** [comment]
pr with updated docs instructions is here #22132

**Pull request merged**